### PR TITLE
Retain semicolon during updates of IndexerDeclaration

### PR DESCRIPTION
**Pull request by GrahamTheCoder**
Spotted this issue while skimming through for similar issues to https://github.com/dotnet/roslyn/pull/68762

Note: I've checked that the test fails prior to the production code change

### CODE REVIEW

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxFactoryTests.cs [hunk 1]
```diff
@@ -83,6 +83,25 @@ public void TestConstructClassWithKindAndNameOnly()
...
+
+            var mutatedDeclaration = initialDeclaration.Update(initialDeclaration.AttributeLists,
+                initialDeclaration.Modifiers,
+                initialDeclaration.Type,
+                initialDeclaration.ExplicitInterfaceSpecifier,
+                initialDeclaration.ThisKeyword,
+                initialDeclaration.ParameterList,
+                initialDeclaration.AccessorList);
+
+            Assert.Contains(comment.ToString(), mutatedDeclaration.SemicolonToken.GetLeadingTrivia().ToFullString());
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxFactoryTests.cs:102]: In general, we prefer doing something like `AssertEx.Equals("text", mutatedDeclaration.SemicolonToken.GetLeadingTrivia().ToFullString())`. It makes it easier to read the test at the cost of a very minor, local duplication, which is worth it imo.

**Review by 333fred — APPROVED** [review]
@dotnet/roslyn-compiler for a second review.

**Review by sharwell — CHANGES_REQUESTED** [review]
I believe this `Update` method is part of converting an expression-bodied indexer (which has a semicolon) to an expanded indexer (which produces error CS1597 for a semicolon after an accessor block).

**Comment by CyrusNajmabadi** [comment]
Update sound preserve all non-provided children.  That's the invariant it provides. If a feature depends on this bug, we should fix the feature (which should be fairly trivial, given that we have to do the same thing for all the other variations the feature works on) :-)

**Comment by sharwell** [comment]
@CyrusNajmabadi That "might" make sense if this change also preserved the `expressionBody`, but it forces that to null as well.

I believe we should drop the semicolon token, but retain its trivia as appended trailing trivia of the updated indexer.

**Comment by GrahamTheCoder** [comment]
This overload is specifically expecting people to use it to pass a non-null accessor list, which is hinted by the non-nullable type it takes, but not hinted in the name or checked anywhere.
I noticed `explicitInterfaceSpecifier` is similarly marked non-null, though for reasons I can't immediately see. Presumably built for a quite specific use case. If there's no known reason, it may be worth changing assuming nullability be actually changed on the public API without too much drama since that would accentuate the one that actually does need to not be null.

I believe this PR would break many common usages, so I'll close it. The remaining issues are:
1. No tests failed when I made this breaking change.
2. This API will do surprising things when passed a null accessorList (which *does* cause a warning) - worth adding a check, or are annotations sufficient?
  * Usually I'd expect passing the existing values from a piece of syntax to an `Update` method to return an equivalent piece of syntax.
3. Some trivia gets ditched - don't bother fixing, worth a quick note in the docs?
  * Unless there are already helper methods for porting trivia around correctly, I don't suggest here is the first place to try fixing that. There are subtleties such as only being allowed one piece of trailing trivia, as well as structured trivia like regions and conditional compilation directives that will be wrong a lot of the time when moved.

I don't have much stake in any of these issues. I just noticed this slight oddity in passing, so I'm happy to leave it there and let focus drift elsewhere

**Pull request closed without merging**